Added b-Zone API Models

Patricijia opened the pull request.

Patricijia committed:
Added b-Zone API Models

Patricijia commented on backend/cmd/web/models/users_model.go:
```diff
@@ -0,0 +1,146 @@
+	"go.mongodb.org/mongo-driver/mongo"
+)
+
+type UserDBModel struct {
```
UsersDBModel* 🥴

Patricijia commented on backend/cmd/web/models/users_model.go:
```diff
@@ -0,0 +1,146 @@
+
+// UsersModel struct for UsersModel
+type UsersModel struct {
+	//
```
And I just realized that we forgot the most important part - how to relate users to their Zones. Either we could store **user ID in the ZonesPlotModel** or the **[]ZonesPlot in the UsersModel**

NeaGogu replied:
I would say store the ZonesPlots ids in the UserModel

Patricijia committed:
Update users_model.go

NeaGogu commented on backend/cmd/web/models/users_model.go:
```diff
@@ -0,0 +1,183 @@
+func NewUsersModelWithDefaults() *UsersModel {
+	this := UsersModel{}
+	return &this
+}
```
What should be the difference between these 2?

NeaGogu commented on backend/cmd/web/models/users_model.go:
```diff
@@ -0,0 +1,183 @@
+		return ret
+	}
+	return *o.Id
+}
```
I don't think we would need this kind of getter. Also returning the zero value of the field in case it does not exist might lead to shooting ourselves in the foot. The type of getter below this (the `Get..Ok`) would be more idiomatic and less error-prone, because you explicitly tell the user (with the returned `bool`  value) that something was not ok, and it's up to the user to decide how to handle that. They could even choose to ignore the `bool` completely and that would be the same as the `GetId()` method.

NeaGogu replied:
Same for all the other methods

Patricijia replied:
So I'll delete all the "get" and "set" methods and leave only GetSomethingOk. Right?

NeaGogu replied:
The "Set" methods are fine, just the "get" ones

NeaGogu commented on backend/cmd/web/models/users_model.go:
```diff
@@ -0,0 +1,183 @@
+	Id *int64 `json:"id,omitempty"`
+	// unique per user
+	Uuid *string `json:"uuid,omitempty"`
+	// zones plots per user
```
Dont store fields with primitive types as pointers. You dont want to return the pointer to those values anyway, cause the user might mutate the field without wanting to when performing operations with that value.

NeaGogu replied:
Same for all the other Models

Patricijia replied:
This is how Bumbal stores stuff in their API. So for all the primitive types e.g. string, int I will delete the pointer, right? But I'll leave "*" for Model types?

NeaGogu replied:
Yeah that sounds good. 
![image](https://user-images.githubusercontent.com/70640237/223983007-a600fe78-54a4-4be4-a207-ad0c3a6b446e.png)
Do you mean the `$int64` thing? I think that just refers to what type of interger it is, not that it's a pointer.

NeaGogu commented on backend/cmd/web/models/users_model.go:
```diff
@@ -0,0 +1,183 @@
+
+// GetIdOk returns a tuple with the Id field value if set, nil otherwise
+// and a boolean to check if the value has been set.
+func (o *UsersModel) GetIdOk() (*int64, bool) {
```
you can even name the return values to be more specific for the user. E.g `(id int64, ok bool)`

NeaGogu commented on backend/cmd/web/models/zone_model.go:
```diff
@@ -0,0 +1,221 @@
+	//Total fuel cost
+	ZoneFuelCost *float64 `json:"zone_fuel_cost,omitempty"`
+	//Total driving time
+	ZoneDrivingTime *int64 `json:"zone_driving_time,omitempty"`
```
same issue as above

NeaGogu commented on backend/cmd/web/models/zone_model.go:
```diff
@@ -0,0 +1,221 @@
+func NewZoneModelWithDefaults() *ZoneModel {
+	this := ZoneModel{}
+	return &this
+}
```
same question as above

NeaGogu commented on backend/cmd/web/models/zone_model.go:
```diff
@@ -0,0 +1,221 @@
+	}
+	return *o.Id
+}
+
```
same comment as above

Patricijia committed:
Updated the models

I deleted all the Get() functions and left only GetOk. I left the pointers because started thinking if we mix and match gonna be a lot of fuss and silly mistakes. Let's try working with pointers if we see that pointers are annoying I can change.

NeaGogu commented:
@Patricijia resolve the conflicts before merging pls. And @VictorVader review pls

Patricijia committed:
Merge branch 'dev' into 59-s1-add-b-zone-api-models

# Conflicts:
#	backend/go.mod
#	backend/go.sum

Patricijia committed:
Resolved conflicts and updated the structure

Patricijia committed:
Updated Models

Patricijia commented:
@NeaGogu pls

NeaGogu commented:
Bruh I'm at the gym 😂😂

NeaGogu commented:
![image](https://user-images.githubusercontent.com/70640237/224421471-adeecf06-1ac8-479e-891f-5337d97dfada.png)
Didn't we decide we're gonna change them to Plot, Zone and ZoneRange? :laughing:
